src/threshold.jl

```diff
@@ -1,30 +1,41 @@
-function threshold_dithering(
-    img::AbstractMatrix{<:Gray}, threshold_map::AbstractMatrix{<:Gray}; to_linear=false
-)::BitMatrix
-    size(img) ≢ size(threshold_map) && throw(
-        DimensionMismatch("Arguments img and threshold_map need to have same dimensions"),
-    )
+abstract type AbstractThresholdDither <: AbstractDither end
 
+function dither(
+    img::AbstractMatrix{<:Gray}, alg::AbstractThresholdDither; to_linear=false
+)::Matrix{Gray{Bool}}
+    tmap = alg(img)
     to_linear && (img = mappedarray(srgb2linear, img))
-    return img .> threshold_map
+    return img .> tmap
 end
 
-function threshold_dithering(
-    img::AbstractMatrix{<:Gray}; threshold::Real=0.5, kwargs...
-)::BitMatrix
-    if threshold < 0 || threshold > 1
-        throw(
-            DomainError(
-                threshold, "Keyword argument threshold needs to be between 0 and 1."
-            ),
-        )
-    end
+"""
+    WhiteNoiseThreshold()
 
-    tmap = fill(Gray{N0f16}(threshold), size(img)) # constant matrix of value threshold
-    return threshold_dithering(img, tmap; kwargs...)
+Use white noise as a threshold map.
+"""
+struct WhiteNoiseThreshold <: AbstractThresholdDither end
+function (alg::WhiteNoiseThreshold)(img)
+    tmap = rand(Gray{N0f16}, size(img)) # threshold map is white noise
+    return tmap
 end
 
-function white_noise_dithering(img::AbstractMatrix{<:Gray}; kwargs...)::BitMatrix
-    tmap = rand(Gray{N0f16}, size(img)) # threshold map is white noise
-    return threshold_dithering(img, tmap; kwargs...)
+"""
+    ConstantThreshold(threshold)
+
+Use a constant threshold map. Defaults to 0.5 if `threshold` isn't specified.
+"""
+struct ConstantThreshold{T<:Real} <: AbstractThresholdDither
+    threshold::T
+
+    function ConstantThreshold(; threshold=0.5)
+        if threshold < 0 || threshold > 1
+            throw(DomainError(threshold, "Threshold needs to be between 0 and 1."))
+        end
+        return new{typeof(threshold)}(threshold)
+    end
+end
+
+function (alg::ConstantThreshold)(img)
+    tmap = fill(Gray{N0f16}(alg.threshold), size(img)) # constant matrix of value threshold
```

Comment:
I just notice this, is `N0f16` deliberately chosen? If so, maybe leave a comment on it? That would be more readable for future potential contributors.

Reply:
Adressed in ceef339